src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderRepository.cs

```diff
@@ -0,0 +1,343 @@
+                        header = row.Value;
+
+                    this.logger.LogTrace("(-)");
+
+                    return header;
+                }
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task PutAsync(List<ProvenBlockHeader> headers, HashHeightPair newTip)  
+        {
+            Guard.NotNull(headers, nameof(headers));
+            Guard.NotNull(newTip, nameof(newTip));            
+
+            if ((this.provenBlockHeaderTip != null) && (newTip.Hash != this.provenBlockHeaderTip.HashPrevBlock))
+            {
+                this.logger.LogTrace("(-)[BLOCKHASH_MISMATCH]");
+                throw new InvalidOperationException("Invalid newTip block hash.");
+            }
+
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}.Count():{1})", nameof(headers), headers.Count());
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    transaction.SynchronizeTables(BlockHashHeightTable, ProvenBlockHeaderTable);
+
+                    this.InsertHeaders(transaction, headers, newTip);
+                    this.SetTip(transaction, newTip);
+
+                    transaction.Commit();
+                }
+
+                this.logger.LogTrace("(-)");
+            });
+
+            return task;
+        }
+
+        /// <inheritdoc />
+        public Task<HashHeightPair> GetTipHashHeightAsync()
+        {
+            Task<HashHeightPair> task = Task.Run(() =>
+            {
+                this.logger.LogTrace("()");
+
+                HashHeightPair tip;
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    tip = this.GetTipHashHeight(transaction);
+                }
+
+                this.logger.LogTrace("(-):'{0}'", tip);
+
+                return tip;
+            });
+
+            return task;
+        }
+
+        /// <summary>
+        /// Delete <see cref="ProvenBlockHeader"/> items.
+        /// </summary>
+        /// <param name="fromBlockHeight">Block height to start range.</param>
+        /// <param name="toBlockHeight">Block height end range.</param>
+        private Task DeleteAsync(int fromBlockHeight, int toBlockHeight)
+        {
+            Task task = Task.Run(() =>
+            {
+                this.logger.LogTrace("({0}:'{1}')", nameof(fromBlockHeight), fromBlockHeight);
+                this.logger.LogTrace("({0}:'{1}')", nameof(toBlockHeight), toBlockHeight);
+
+                using (DBreeze.Transactions.Transaction transaction = this.dbreeze.GetTransaction())
+                {
+                    transaction.SynchronizeTables(BlockHashHeightTable, ProvenBlockHeaderTable);
+
+                    transaction.ValuesLazyLoadingIsOn = false;
+
+                    for (int i = fromBlockHeight; i <= toBlockHeight; i++)
```

Comment:
Should this not also up[date the tip?
What about making sure sequentialness is kept?